src/dma/mod.rs

```diff
@@ -0,0 +1,799 @@
+    FirstBuffer,
+    /// The second buffer (m1ar) is in use.
+    DoubleBuffer,
+}
+
+impl Not for CurrentBuffer {
+    type Output = CurrentBuffer;
+
+    fn not(self) -> Self::Output {
+        if self == CurrentBuffer::FirstBuffer {
+            CurrentBuffer::DoubleBuffer
+        } else {
+            CurrentBuffer::FirstBuffer
+        }
+    }
+}
+
+/// Contains types related to DMA configuration.
+pub mod config {
+    use super::Bits;
+
+    /// Priority of the DMA stream, defaults to `Medium`. If two requests have
+    /// the same software priority level, the stream with the lower number takes
+    /// priority over the stream with the higher number. For example, Stream 2
+    /// takes priority over Stream 4.
+    #[derive(Debug, Clone, Copy)]
+    pub enum Priority {
+        /// Low priority.
+        Low,
+        /// Medium priority.
+        Medium,
+        /// High priority.
+        High,
+        /// Very high priority.
+        VeryHigh,
+    }
+    impl Default for Priority {
+        fn default() -> Self {
+            Priority::Medium
+        }
+    }
+
+    impl Bits<u8> for Priority {
+        fn bits(self) -> u8 {
+            match self {
+                Priority::Low => 0,
+                Priority::Medium => 1,
+                Priority::High => 2,
+                Priority::VeryHigh => 3,
+            }
+        }
+    }
+
+    /// The level to fill the fifo to before performing the transaction.
+    #[derive(Debug, Clone, Copy)]
+    pub enum FifoThreshold {
+        /// 1/4 full.
+        QuarterFull,
+        /// 1/2 full.
+        HalfFull,
+        /// 3/4 full.
+        ThreeQuarterFull,
+        /// Full.
+        Full,
+    }
+
+    impl Bits<u8> for FifoThreshold {
+        fn bits(self) -> u8 {
+            match self {
+                FifoThreshold::QuarterFull => 0,
+                FifoThreshold::HalfFull => 1,
+                FifoThreshold::ThreeQuarterFull => 2,
+                FifoThreshold::Full => 3,
+            }
+        }
+    }
+
+    /// How burst transfers are done, requires fifo enabled. Check datasheet for
+    /// valid combinations.
+    #[derive(Debug, Clone, Copy)]
+    pub enum BurstMode {
+        /// Single transfer, no burst.
+        NoBurst,
+        /// Burst transfer of 4 beats.
+        Burst4,
+        /// Burst transfer of 8 beats.
+        Burst8,
+        /// Burst transfer of 16 beats.
+        Burst16,
+    }
+
+    impl Bits<u8> for BurstMode {
+        fn bits(self) -> u8 {
+            match self {
+                BurstMode::NoBurst => 0,
+                BurstMode::Burst4 => 1,
+                BurstMode::Burst8 => 2,
+                BurstMode::Burst16 => 3,
+            }
+        }
+    }
+}
+
+/// DMA Transfer.
+pub struct Transfer<STREAM, PERIPHERAL, DIR, BUF>
+where
+    STREAM: Stream,
+    PERIPHERAL: TargetAddress<DIR>,
+    DIR: Direction,
+    BUF: WriteBuffer<Word = <PERIPHERAL as TargetAddress<DIR>>::MemSize>
```

Comment:
I believe the implication here is that we can't provide a constant buffer in a MemoryToPeripheral transfer since we have to provide a mutable ref. This is not desirable if we are doing e.g. a circular transfer of some constant code-word